src/zerocoin_v3.cpp

```diff
@@ -600,29 +600,26 @@ bool CSigmaState::AddMintsToStateAndBlockIndex(
         CBlockIndex *index,
         const CBlock* pblock) {
 
-    std::unordered_map<sigma::CoinDenomination, size_t> blockDenomCounts;
+    std::unordered_map<sigma::CoinDenomination, std::vector<sigma::PublicCoin>> blockDenomMints;
     for (const auto& mint : pblock->sigmaTxInfo->mints) {
-        blockDenomCounts[mint.getDenomination()]++;
+        blockDenomMints[mint.getDenomination()].push_back(mint);
     }
 
-    for (const auto& pubCoin : pblock->sigmaTxInfo->mints) {
-        sigma::CoinDenomination denomination = pubCoin.getDenomination();
+    for (const auto& it : blockDenomMints) {
+        const sigma::CoinDenomination denomination = it.first;
+        const std::vector<sigma::PublicCoin>& mintsWithThisDenom = it.second;
 
         if (latestCoinIds[denomination] < 1)
             latestCoinIds[denomination] = 1;
-        int	mintCoinGroupId = latestCoinIds[denomination];
+        auto mintCoinGroupId = latestCoinIds[denomination];
+
 
-        // ZC_SPEND_V3_COINSPERID = 15.000, yet the actual limit of coins per accumlator is 16.000.
-        // We need to cut at 15.000, such that we always have enough space for new mints. Mints for
-        // each block will end up in the same accumulator.
         SigmaCoinGroupInfo &coinGroup = coinGroups[make_pair(denomination, mintCoinGroupId)];
         int coinsPerId = ZC_SPEND_V3_COINSPERID;
         int coinsPerIdLimit = ZC_SPEND_V3_COINSPERID_LIMIT;
-        if ((coinGroup.nCoins < coinsPerId // there's still space in the accumulator
-            && coinGroup.nCoins + blockDenomCounts[denomination] < coinsPerIdLimit)
-            || coinGroup.lastBlock == index // or we have already placed some coins from current block.
-            ) {
-            if (coinGroup.nCoins++ == 0) {
+        if (coinGroup.nCoins < coinsPerId // there's still space in the accumulator
+            && coinGroup.nCoins + mintsWithThisDenom.size() <= coinsPerIdLimit) {
```

Comment:
Seem like this one will cause signesss comparison mismatch.

Reply:
fixed